### PR TITLE
UI: Add system tray capability

### PR DESCRIPTION
**Pull request by cg2121**
This adds a system tray icon when enabled. There is also a option to minimize to the system tray when the app is started.

### CODE REVIEW

**Comment by ONSBalder** [comment]
Can one of the admins verify this patch?

**Comment by reboot** [comment]
I would disable the "Minimize to system try when started" option if "Enable system tray icon" is not enabled. It makes no sense to minimize to the tray, when you have no tray icon.

If you close OBS using the exit menu entry from the tray icon while the window is hidden then the window position is lost and the window is not placed at the original position on the next start.

**Comment by cg2121** [comment]
All of those issues are fixed now (hopefully).

File: obs/window-basic-main.cpp [hunk 1]
```diff
@@ -4403,3 +4439,90 @@ void OBSBasic::on_actionLockPreview_triggered()
...
+	trayMenu->addAction(sysTrayRecord);
+	trayMenu->addAction(showHide);
+	trayMenu->addAction(exit);
+	trayIcon->setContextMenu(trayMenu);
+}
+
+void OBSBasic::SystemTrayExit()
+{
+	QTimer::singleShot(250, this, SLOT(show()));
+	close();
```
Comment (reboot) [inline obs/window-basic-main.cpp]: The window position is still lost for me, when i close OBS using the exit menu entry. I'm not sure if this code makes sense. You try to make the window visible with a 250ms delay, but then you immediately close it. Maybe a better solution would be to remove the geometry saving from the destructor and execute it every time the window is hidden, which should happen on normal exit and when using "Hide" from the tray icon. Then you can close the hidden window without messing up the saved geometry.

**Comment by cg2121** [comment]
See if that works for you now.

**Comment by cg2121** [comment]
Just realized that I put the save function in the wrong place. It is in the right place now.

**Comment by reboot** [comment]
It works now, but I think it would have been enough to just use isVisible() instead of adding the saveGeom property.

Hope you don't mind all this criticism, getting stuff to work correctly is of course required, improving the code is more for making it more maintainable in the future, and maybe some development experience for future stuff.

**Comment by cg2121** [comment]
I have now added a notification if the stream fails to connect if the main window is hidden. I also disabled the stream confirmation boxes if main window is hidden. Else if these boxes would show when not visible and you click okay a segmentation fault would happen.

**Comment by cg2121** [comment]
I have now made it now that the system tray icon is only loaded into memory when it is needed, and deleted when it is not.

**Comment by jp9000** [comment]
Finally took a look at this, I love it.  However, one thing:  I think it shouldn't use tray popup notifications.  Popup notifications should be used for things the user has no control over.

Instead, I feel that the icon should change, similar to how OBS1 did it, where it had a little red dot-looking thing on the tray icon indicating that the program is active.  One of the things people have requested the most is a notification like that.  Would be a bit more ideal in my opinion, would that be okay?  I have some free time and can help out with it if needed, if it's okay I'll just modify it to do that.  Besides I think I may have to size-down the icon to look a bit better as a tray icon.

Also, you have a fair amount of trailing whitespace.

**Comment by cg2121** [comment]
I will be able to work on this again later today.

**Comment by cg2121** [comment]
I have now added a indicator to the icon when streaming or recording is active. When they are active a red dot appears, like in OBS Classic.

**Comment by cg2121** [comment]
I am done with this now, so you can merge it whenever you want, or if anything needs changing I will still make those changes.

**Comment by cg2121** [comment]
Now I am finally done. I forgot to add a toggle to show and hide when clicking the icon.

**Comment by jp9000** [comment]
Okay, I modified the PR a slight bit:  I added tray notification messages for disconnection/reconnection, and refactored a bit of my own code in window-basic-main.cpp because I noticed you had to repeat a lot of code for when streaming starts/stops.

**Pull request closed without merging**